saleor/graphql/core/mutations.py

```diff
@@ -65,146 +47,159 @@ def _update_mutation_arguments_and_fields(cls, arguments, fields):
         cls._meta.fields.update(fields)
 
 
-class ModelFormMutationOptions(MutationOptions):
-    form_class = None
-    return_field_name = None
-
-
-class ModelDeleteMutationOptions(MutationOptions):
-    model = None
-    return_field_name = None
-
-
-class ModelDeleteMutation(BaseMutation):
-
+class ModelMutation(BaseMutation):
     class Meta:
         abstract = True
 
     @classmethod
     def __init_subclass_with_meta__(
-            cls, arguments=None, model=None, return_field_name=None,
-            _meta=None, **options):
+            cls, arguments=None, model=None, permissions=None, _meta=None,
+            **options):
         if not model:
-            raise ImproperlyConfigured(
-                'model is required for ModelDeleteMutation')
-
-        _meta = ModelDeleteMutationOptions(cls)
-
-        if not return_field_name:
-            return_field_name = get_model_name(model)
+            raise ImproperlyConfigured('model is required for ModelMutation')
+        _meta = ModelMutationOptions(cls)
+        return_field_name = get_model_name(model)
         if arguments is None:
             arguments = {}
-        arguments.update({'id': graphene.ID()})
         fields = get_output_fields(model, return_field_name)
-
         _meta.model = model
         _meta.return_field_name = return_field_name
-
         super().__init_subclass_with_meta__(_meta=_meta, **options)
-
-        # Update mutation's arguments and fields
         cls._update_mutation_arguments_and_fields(
             arguments=arguments, fields=fields)
 
     @classmethod
-    def mutate(cls, root, info, id, **kwargs):
-        model = cls._meta.model
-        model_type = registry.get_type_for_model(model)
-        instance = get_node(info, id, only_type=model_type)
-        instance.delete()
-        field_name = cls._meta.return_field_name
-        kwargs = {field_name: instance}
-        return cls(**kwargs)
+    def add_error(cls, errors, field, message):
+        errors.append(Error(field=field, message=message))
+        return errors
 
+    @classmethod
+    def _check_type(cls, field, target_type):
+        if hasattr(field.type, 'of_type'):
+            return field.type.of_type == target_type
+        return field.type == target_type
 
-class ModelFormMutation(BaseMutation):
-
-    class Meta:
-        abstract = True
+    @classmethod
+    def _clean_input(cls, info, instance, input, errors):
```

Comment:
I think argument `errors` should be dropped from here. It is not being used here apart from being returned in the same way as while called.
For me it is a bit misleading, for example in `mutate` method below, `errors = []` is being initialised, then this function is called, which does nothing and then there is a call to `_clean_instance` which checks for errors.

Reply:
The idea is that at any step of the mutation execution you are able to add your own errors. It's not used in the original implementation, but if you customized `clean_input` and you had some custom conditions to validate the data, you'd be able to do this.

Reply:
Edit: at any "validation" step so `clean_input` or `clean_instance`.